File: src/decoders/notebook.ts
import { Notebook, NotebookAbsence, NotebookDelay, NotebookObservation, NotebookPrecautionaryMeasure, NotebookPunishment, SessionHandle } from "~/models";
import { decodeNotebookAbsence } from "./notebook-absence";
import { decodeNotebookDelay } from "./notebook-delay";
import { decodeNotebookPunishment } from "./notebook-punishment";
import { decodeNotebookObservation } from "./notebook-observation";
import { decodeNotebookPrecautionaryMeasure } from "./notebook-precautionary-measure";

export const decodeNotebook = (notebook: any, session: SessionHandle): Notebook => {
  const absences: NotebookAbsence[] = [];
  const delays: NotebookDelay[] = [];
  const punishments: NotebookPunishment[] = [];
  const observations: NotebookObservation[] = [];
  const precautionaryMeasures: NotebookPrecautionaryMeasure[] = [];

  for (const item of notebook.listeAbsences.V) {
    switch (item.G) {
      case 13: // Absence
        absences.push(decodeNotebookAbsence(item));
        break;
      case 14: // Delay
        delays.push(decodeNotebookDelay(item));
        break;
      case 41: // Punishment
        punishments.push(decodeNotebookPunishment(item, session));
        break;
      case 46: // ObservationProfesseurEleve
        observations.push(decodeNotebookObservation(item));
        break;
      case 72: // PrecautionaryMeasure
        precautionaryMeasures.push(decodeNotebookPrecautionaryMeasure(item, session));
        break;
    }
  }

  // TODO: .Matieres
  // TODO: .listeRecapitulatifs
  // NOTE: not sure if .autorisations is needed, for now.
  // NOTE: not sure what .nbMaxJoursDeclarationAbsence and .listeSanctionUtilisateur are about, for now.

  return {
    absences,
    delays,
    punishments,
    observations,
    precautionaryMeasures
  };
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/decoders/notebook.kt
package ink.literate.pawnote.decoders

import ink.literate.pawnote.models.*
import kotlinx.serialization.json.*

fun decodeNotebook(notebook: JsonObject, session: SessionHandle): Notebook {
  val absences: MutableList<NotebookAbsence> = mutableListOf()
  val delays: MutableList<NotebookDelay> = mutableListOf()
  val punishments: MutableList<NotebookPunishment> = mutableListOf()
  val observations: MutableList<NotebookObservation> = mutableListOf()
  val precautionaryMeasures: MutableList<NotebookPrecautionaryMeasure> = mutableListOf()

  for (item in notebook["listeAbsences"]!!.jsonObject["V"]!!.jsonArray.map { it.jsonObject }) {
    when (item["G"]!!.jsonPrimitive.int) {
      13 -> absences.add(decodeNotebookAbsence(item))
      14 -> delays.add(decodeNotebookDelay(item))
      41 -> punishments.add(decodeNotebookPunishment(item, session))
      46 -> observations.add(decodeNotebookObservation(item))
      72 -> precautionaryMeasures.add(decodeNotebookPrecautionaryMeasure(item, session))
    }
  }

  // TODO: .Matieres
  // TODO: .listeRecapitulatifs
  // NOTE: not sure if .autorisations is needed, for now.
  // NOTE: not sure what .nbMaxJoursDeclarationAbsence and .listeSanctionUtilisateur are about, for
  // now.

  return Notebook(
      absences = absences,
      delays = delays,
      punishments = punishments,
      observations = observations,
      precautionaryMeasures = precautionaryMeasures)
}
